cpp/src/arrow/compute/kernels/take.cc

```diff
@@ -0,0 +1,351 @@
+    auto index_data = GetValues<int32_t>(indices, 1);
+    const uint8_t* input_data = input.buffers[1]->data();
+    uint8_t* out_data = output->buffers[1]->mutable_data();
+    internal::BitmapWriter bit_writer(out_data, output->offset, indices.length);
+    for (int64_t i = 0; i < indices.length; ++i) {
+      if (BitUtil::GetBit(input_data, *index_data++)) {
+        bit_writer.Set();
+      } else {
+        bit_writer.Clear();
+      }
+      bit_writer.Next();
+    }
+    bit_writer.Finish();
+  }
+};
+
+template <typename I>
+struct TakeFunctor<NullType, I, enable_if_index_valid<I>> {
+  void operator()(FunctionContext* ctx, const TakeOptions& options,
+                  const ArrayData& input, const ArrayData& indices, ArrayData* output) {}
+};
+
+// ----------------------------------------------------------------------
+
+typedef std::function<void(FunctionContext*,
+                           const TakeOptions& options,
+                           const ArrayData&,
+                           const ArrayData&,
+                           ArrayData*)>
+    TakeFunction;
+
+static Status AllocateIfNotPreallocated(FunctionContext* ctx,
+                                        const int64_t length,
+                                        ArrayData* out) {
+  if (out->buffers.size() == 2) { // pre-allocated
+    return Status::OK();
+  } else {
+    DCHECK_EQ(0, out->buffers.size());
+  }
+
+  std::shared_ptr<Buffer> validity_bitmap;
+  int64_t bitmap_size = BitUtil::BytesForBits(length);
+  RETURN_NOT_OK(ctx->Allocate(bitmap_size, &validity_bitmap));
+  if (out->type->id() == Type::NA) {
+    out->null_count = length;
+    memset(validity_bitmap->mutable_data(), 0, bitmap_size);
+  } else {
+    out->null_count = 0;
+    memset(validity_bitmap->mutable_data(), 0xFF, bitmap_size); // preset all to valid
+  }
+
+  const Type::type type_id = out->type->id();
+  if (!(is_primitive(type_id) || type_id == Type::FIXED_SIZE_BINARY ||
+      type_id == Type::DECIMAL)) {
+    std::stringstream ss;
+    ss << "Cannot pre-allocate memory for type: " << out->type->ToString();
+    return Status::NotImplemented(ss.str());
+  }
+
+  std::shared_ptr<Buffer> out_data;
+  if (type_id != Type::NA) {
+    const auto& fw_type = static_cast<const FixedWidthType&>(*out->type);
+    int bit_width = fw_type.bit_width();
+    int64_t buffer_size = 0;
+    if (bit_width == 1) {
+      buffer_size = BitUtil::BytesForBits(length);
+    } else if (bit_width % 8 == 0) {
+      buffer_size = length * fw_type.bit_width() / 8;
+    } else {
+      DCHECK(false);
+    }
+    RETURN_NOT_OK(ctx->Allocate(buffer_size, &out_data));
+    memset(out_data->mutable_data(), 0, buffer_size);
+  }
+
+  out->buffers.push_back(validity_bitmap);
+  out->buffers.push_back(out_data);
+
+  return Status::OK();
+}
+
+template <typename IndexType>
+class TakeKernel : public UnaryKernel {};
+
+template <>
+class TakeKernel<Int32Type> : public UnaryKernel {
```

Comment:
Why `Int32Type`, this seems fishy as it doesn't seem used in the rest of the code.